File: packages/core/src/sectors/nigeria_insurance.ts
/**
 * NigeriaInsuranceCompliance — TypeScript sector pack.
 *
 * Composes: NAICOM 2021/2023, NDPA 2023, NFIU/MLPPA 2022,
 *           OWASP Universal (PII, injection, tools, approval).
 *
 * Covers AI agents in: claims processing, automated underwriting,
 *                       fraud detection, policy management.
 */

import { Comply54Engine } from "../engine.js";
import { ReceiptSigner } from "../receipts.js";
import { PACK_VERSIONS } from "../packs/versions.js";
import { evaluateNAICOM, evaluateNDPA, evaluateNfiu } from "../packs/nigeria.js";
import {
  evaluatePiiLeakage,
  evaluatePromptInjection,
  evaluateToolPermissions,
  evaluateHumanApproval,
} from "../packs/universal.js";
import type { ComplianceCertificate, ComplianceResult, EvaluationInput } from "../types.js";

const PACKS = [
  evaluateNAICOM,
  evaluateNDPA,
  evaluateNfiu,
  evaluatePiiLeakage,
  evaluatePromptInjection,
  evaluateToolPermissions,
  evaluateHumanApproval,
];

const REGULATIONS = [
  "Insurance Act 2003 (Cap I17 LFN 2004)",
  "NAICOM Operational Guidelines 2021",
  "NAICOM Market Conduct Guidelines 2023",
  "Nigeria Data Protection Act 2023",
  "MLPPA 2022 / NFIU AML Guidelines",
  "OWASP Agentic AI",
];

export class NigeriaInsuranceCompliance {
  readonly name = "Nigeria Insurance Compliance";
  readonly jurisdictions = ["NG"];
  readonly regulations = REGULATIONS;

  private readonly engine: Comply54Engine;
  private readonly signer?: ReceiptSigner;

  constructor(options: { strictMode?: boolean; signingKey?: string } = {}) {
    this.engine = new Comply54Engine(PACKS, options);
    if (options.signingKey) {
      this.signer = new ReceiptSigner(options.signingKey);
    }
  }

  check(
    action: string,
    params?: Record<string, unknown>,
    output?: string,
    context?: Record<string, unknown>
  ): ComplianceResult {
    return this.engine.check(action, params, output, context);
  }

  /** Like `check()` but returns a `Promise<ComplianceResult>` with `receiptToken` populated. */
  async checkSigned(
    action: string,
    params?: Record<string, unknown>,
    output?: string,
    context?: Record<string, unknown>
  ): Promise<ComplianceResult> {
    if (!this.signer) {
      throw new Error("checkSigned() requires signingKey to be set in the constructor");
    }
    const result = this.check(action, params, output, context);
    const packVersions = Object.fromEntries(
      result.decisions.map((d) => [d.pack, PACK_VERSIONS[d.pack] ?? "unknown"])
    );
    const token = await this.signer.sign(result, action, params ?? {}, output ?? "", context ?? {}, packVersions);
    return { ...result, receiptToken: token };
  }

  evaluate(input: EvaluationInput): ComplianceResult {
    return this.engine.evaluate(input);
  }

  async certificate(
    action: string,
    params?: Record<string, unknown>,
    output?: string,
    context?: Record<string, unknown>
  ): Promise<ComplianceCertificate> {
    return this.engine.certificate(
      { action, params, output, context },
      {
        sectorPack: this.name,
        jurisdictions: this.jurisdictions,
        regulations: this.regulations,
      }
    );
  }
}

File: comply54/sectors/nigeria_insurance.py
"""
comply54.sectors.nigeria_insurance
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Nigerian insurance compliance sector pack.

Composes:
  - NIIRA 2025            — Nigerian Insurance Industry Reform Act 2025 (signed August 2025).
                            Replaces Insurance Act 2003 (Cap I17 LFN 2004). Key provisions:
                            §210 mandatory 60-day claims settlement, §212 Policyholders'
                            Protection Fund, Part V market conduct and anti-discrimination rules.
  - NAICOM 2021/2023      — Claims adjudication thresholds, underwriting controls, fraud
                            escalation (remain in force under NIIRA 2025 pending new guidelines)
  - NDPA 2023             — Personal data including health/biometric used in underwriting
  - NFIU AML              — Large premium / claim AML reporting duty
  - OWASP Universal       — PII leakage, prompt injection, tool permissions, human approval

Covers AI agents operating in:
  - Claims processing systems
  - Automated underwriting engines
  - Fraud detection platforms
  - Policy management systems

Usage:
    from comply54.sectors import NigeriaInsuranceCompliance

    compliance = NigeriaInsuranceCompliance()
    result = compliance.check(
        action="deny_claim",
        params={"claim_id": "CLM-001", "claim_amount": 750000},
        context={"human_adjuster_assigned": False},
    )
    if result.blocked:
        raise ValueError(result.primary_violation.messages[0])
"""

from __future__ import annotations

from ..core.packs import (
    HUMAN_APPROVAL,
    NAICOM,
    NDPA,
    NFIU_AML,
    PII_LEAKAGE,
    PROMPT_INJECTION,
    TOOL_PERMISSIONS,
)
from ._base import SectorCompliance


class NigeriaInsuranceCompliance(SectorCompliance):
    """
    Full Nigerian insurance compliance pack.

    Covers Nigerian Insurance Industry Reform Act 2025 (NIIRA 2025) §§210, 212, Part V,
    NAICOM Operational Guidelines 2021 (Guidelines 12, 15, 18), NAICOM Market Conduct
    Guidelines 2023 (Rules 6, 11), NFIU AML Guidelines (insurance sector), NDPA 2023,
    and OWASP Agentic AI.

    NIIRA 2025 replaced the Insurance Act 2003 (Cap I17 LFN 2004) in August 2025.
    """

    name = "Nigeria Insurance Compliance"
    jurisdictions = ["NG"]
    regulations = [
        "Nigerian Insurance Industry Reform Act 2025 (NIIRA 2025)",
        "NAICOM Operational Guidelines 2021",
        "NAICOM Market Conduct Guidelines 2023",
        "Nigeria Data Protection Act 2023",
        "MLPPA 2022 / NFIU AML Guidelines",
        "OWASP Agentic AI",
    ]

    def __init__(
        self,
        strict_mode: bool = False,
        signing_key: "bytes | str | None" = None,
    ) -> None:
        super().__init__(
            packs=[
                NAICOM,
                NDPA,
                NFIU_AML,
                PII_LEAKAGE,
                PROMPT_INJECTION,
                TOOL_PERMISSIONS,
                HUMAN_APPROVAL,
            ],
            strict_mode=strict_mode,
            signing_key=signing_key,
        )
